### DISCUSSION TITLE
Feature suggestion: Auto generated tuple members

### DISCUSSION THREAD

**Discussion by taori** [dotnet/csharplang#3665] [Language Ideas]
When writing unit tests if often find myself generating immense bulk of code involving mocks + instances of objects to reduce that amount of bulk. This then results in methods like the following:

![grafik](https://user-images.githubusercontent.com/5545184/87209778-73a1cd00-c313-11ea-8661-50c06818b087.png)

Ideally, since i think this scenario is not that rare i would love if it were possible that i could instead write my code like the following:

    public tuple CreateMock()
		{
			tuple var loaderMock = new Mock<ILoader>();
			tuple var choiceManagerMock = new Mock<IUserChoiceManager>();
			tuple var agentSettingsProviderMock = new Mock<IAgentSettingsProvider>();
			tuple var dialogServiceMock = new Mock<IDialogService>();
			tuple var serviceProviderMock = new Mock<IServiceProvider>();
			tuple var trayIconApiClientFactoryMock = new Mock<ITrayIconApiClientFactory>();
			tuple var windowServiceMock = new Mock<IWindowService>();
			ISettingsTab[] tabs = new ISettingsTab[]
			{
				new CustomerInformationViewModel(trayIconApiClientFactoryMock.Object, choiceManagerMock.Object),
				new NetworkSettingsViewModel(agentSettingsProviderMock.Object, dialogServiceMock.Object), 
			};

			tuple var model = new SettingsWindowViewModel(
				loaderMock.Object, 
				choiceManagerMock.Object, 
				dialogServiceMock.Object, 
				serviceProviderMock.Object, 
				tabs, 
				windowServiceMock.Object);

			return tuple;
		}

Obviously this would greatly reduce the amount of boilerplate required to pass on the values i want to generate

**Comment by masonwheeler** [comment]
There's a language feature that really helps with this.  It's been around since version 1.

ValueTuple is cool, but it's not a golden hammer.  If you're in a situation where it doesn't work particularly well, *don't use it there.*  Just create a struct or class type instead.

**Comment by HaloFour** [comment]
This uses a different keyword and limits to tuples but otherwise it is the same as the following:

#1484

**Comment by taori** [comment]
> 
> 
> There's a language feature that really helps with this. It's been around since version 1.
> 
> ValueTuple is cool, but it's not a golden hammer. If you're in a situation where it doesn't work particularly well, _don't use it there._ Just create a struct or class type instead.

That is exactly the point why i want to use tuples. I want to reduce boilerplate. creating a struct to capture my variables would do the exact oppsite of what i want to accomplish.

**Comment by mikernet** [comment]
`var` return types would handle this well but I believe the opinion on those is that they would be too fragile.

**Comment by taori** [comment]
I don't see how that would make it fragile, but ok. I just think it could reduce boilerplate a lot whereever it is used. Though i don't use tuples in many places, wherever the returned members will be more than just a couple it would significantly cut down boilerplate. In the given sample it would be 9 lines less

**Comment by HaloFour** [comment]
The "fragility" comes from code changes accidentally impacting the signature of the method.  The referenced issue goes into further issues as to why something like this would seriously complicate the Roslyn compiler due to the assumptions that it makes and what impact that would have on performance.  Even limited to tuples the compiler would still have to infer every generic type argument in the signature.

**Comment by taori** [comment]
> 
> 
> The "fragility" comes from code changes accidentally impacting the signature of the method. The referenced issue goes into further issues as to why something like this would seriously complicate the Roslyn compiler due to the assumptions that it makes and what impact that would have on performance. Even limited to tuples the compiler would still have to infer every generic type argument in the signature.

Hm i suppose if there are performance implications tied to it the feature would not be worth it. I suppose i'll have a look whether a source generator would allow for custom syntax features then, since  that would also suit my needs if that is possible. Thanks for the information i guess

**Comment by jnm2** [comment]
@taori This is what I've been doing:

```cs
using NSubstitute;
using NUnit.Framework;
using Shouldly;

public static class SomeTests
{
    [Test]
    public static void Some_test()
    {
        var fixture = new Fixture();
        fixture.WindowService.Foo().Returns(42);
        
        fixture.Model.Bar().ShouldBe(42);
    }
    
    [Test]
    public static void Another_test()
    {
        var fixture = new Fixture();
        // ...
    }
    
    private sealed class Fixture // Sometimes IDisposable if useful
    {
        public ILoader Loader { get; } = Substitute.For<ILoader>();
        public IUserChoiceManager ChoiceManager { get; } = Substitute.For<IUserChoiceManager>();
        public IAgentSettingsProvider AgentSettingsProvider { get; } = Substitute.For<IAgentSettingsProvider>();
        public IDialogService DialogService { get; } = Substitute.For<IDialogService>();
        public IServiceProvider ServiceProvider { get; } = Substitute.For<IServiceProvider>();
        public ITrayIconApiClientFactory TrayIconApiClientFactory { get; } = Substitute.For<ITrayIconApiClientFactory>();
        public IWindowService WindowService { get; } = Substitute.For<IWindowService>();
        
        public SettingsWindowViewModel Model { get; }
        
        // Sometimes helper methods or settable properties if useful
        
        public Fixture() // Sometimes optional params if useful
        {
            Model = new SettingsWindowViewModel(
                Loader, 
                ChoiceManager, 
                DialogService, 
                ServiceProvider, 
                tabs: new ISettingsTab[]
                {
                    new CustomerInformationViewModel(TrayIconApiClientFactory, ChoiceManager),
                    new NetworkSettingsViewModel(AgentSettingsProvider, DialogService), 
                }, 
                WindowService);
        }
    }
}
```
